src/backend/core/services/opensearch.py

```diff
@@ -0,0 +1,357 @@
+
+    logger.info("Performing hybrid search with embedding: %s", q)
+    return {
+        "hybrid": {
+            "queries": [
+                {
+                    "bool": {
+                        "must": {
+                            "multi_match": {
+                                "query": q,
+                                # Give title more importance over content by a power of 3
+                                "fields": ["title.text^3", "content"],
+                            }
+                        },
+                        "filter": filter_,
+                    }
+                },
+                {
+                    "bool": {
+                        "must": {
+                            "knn": {
+                                "embedding": {
+                                    "vector": embedding,
+                                    "k": k,
+                                }
+                            }
+                        },
+                        "filter": filter_,
+                    }
+                },
+            ]
+        }
+    }
+
+
+def get_filter(reach, visited, user_sub, groups):
+    """Build OpenSearch filter"""
+    filters = [
+        {"term": {"is_active": True}},  # filter out inactive documents
+        # Access control filters
+        {
+            "bool": {
+                "should": [
+                    # Public or authenticated (not restricted)
+                    {
+                        "bool": {
+                            "must_not": {
+                                "term": {enums.REACH: enums.ReachEnum.RESTRICTED},
+                            },
+                            "must": {
+                                "terms": {"_id": sorted(visited)},
+                            },
+                        }
+                    },
+                    # Restricted: either user or group must match
+                    {"term": {enums.USERS: user_sub}},
+                    {"terms": {enums.GROUPS: groups}},
+                ],
+                "minimum_should_match": 1,
+            }
+        },
+    ]
+
+    # Optional reach filter
+    if reach is not None:
+        filters.append({"term": {enums.REACH: reach}})
+
+    return filters
+
+
+def get_sort(query_keys, order_by, order_direction):
+    """Build OpenSearch sort clause"""
+    # Add sorting logic based on relevance or specified field
+    if "hybrid" in query_keys:
+        # sorting by other field than "_score" is not supported in hybird search
+        # see: https://github.com/opensearch-project/neural-search/issues/866
+        return {"_score": {"order": order_direction}}
+    if order_by == enums.RELEVANCE:
+        return {"_score": {"order": order_direction}}
+
+    return {order_by: {"order": order_direction}}
+
+
+def get_params(query_keys):
+    """Build OpenSearch search parameters"""
+    if "hybrid" in query_keys:
+        ensure_search_pipeline_exists(HYBRID_SEARCH_PIPELINE_ID)
```

Comment:
Maybe this should be done once, because even if this is not a problem, I think we can prevent calling OpenSearch for nothing at each request.

Several solutions : 
 - create the pipeline in the application `.ready()` method : the pipeline is created at startup
 - create the pipeline in a Django migration : the pipeline is created before the application is deployed, only once and we don't expect the pipeline to be removed + you may create a "management command" which allows to create the pipeline manually in case of emergency